src/main/core/support/units.rs

```diff
@@ -0,0 +1,703 @@
+            Self::Mebi => write!(f, "Mi"),
+            Self::Giga => write!(f, "G"),
+            Self::Gibi => write!(f, "Gi"),
+            Self::Tera => write!(f, "T"),
+            Self::Tebi => write!(f, "Ti"),
+            Self::Base => Ok(()),
+        }
+    }
+}
+
+impl Prefixes for SiPrefixes {
+    fn relative_magnitude(&self) -> u128 {
+        match self {
+            Self::Nano => 1,
+            Self::Micro => 1_000,
+            Self::Milli => 1_000_000,
+            Self::Base => 1_000_000_000,
+            Self::Kilo => 1_000_000_000_000,
+            Self::Kibi => 1_000_000_000 * (1 << 10),
+            Self::Mega => 1_000_000_000_000_000,
+            Self::Mebi => 1_000_000_000 * (1 << 20),
+            Self::Giga => 1_000_000_000_000_000_000,
+            Self::Gibi => 1_000_000_000 * (1 << 30),
+            Self::Tera => 1_000_000_000_000_000_000_000,
+            Self::Tebi => 1_000_000_000 * (1 << 40),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum SiPrefixesUpper {
+    Base,
+    Kilo,
+    Kibi,
+    Mega,
+    Mebi,
+    Giga,
+    Gibi,
+    Tera,
+    Tebi,
+}
+
+impl Default for SiPrefixesUpper {
+    fn default() -> Self {
+        Self::Base
+    }
+}
+
+impl FromStr for SiPrefixesUpper {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "K" | "kilo" => Ok(Self::Kilo),
+            "Ki" | "kibi" => Ok(Self::Kibi),
+            "M" | "mega" => Ok(Self::Mega),
+            "Mi" | "mebi" => Ok(Self::Mebi),
+            "G" | "giga" => Ok(Self::Giga),
+            "Gi" | "gibi" => Ok(Self::Gibi),
+            "T" | "tera" => Ok(Self::Tera),
+            "Ti" | "tebi" => Ok(Self::Tebi),
+            _ => Err("Unit prefix was not one of (K|kilo|Ki|kibi|M|mega|Mi|mebi\
+                |G|giga|Gi|gibi|T|tera|Ti|tebi)"
+                .to_string()),
+        }
+    }
+}
+
+impl fmt::Display for SiPrefixesUpper {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Kilo => write!(f, "K"),
+            Self::Kibi => write!(f, "Ki"),
+            Self::Mega => write!(f, "M"),
+            Self::Mebi => write!(f, "Mi"),
+            Self::Giga => write!(f, "G"),
+            Self::Gibi => write!(f, "Gi"),
+            Self::Tera => write!(f, "T"),
+            Self::Tebi => write!(f, "Ti"),
+            Self::Base => Ok(()),
+        }
+    }
+}
+
+impl Prefixes for SiPrefixesUpper {
+    fn relative_magnitude(&self) -> u128 {
+        match self {
+            Self::Base => 1,
+            Self::Kilo => 1_000,
+            Self::Kibi => 1 << 10,
+            Self::Mega => 1_000_000,
+            Self::Mebi => 1 << 20,
+            Self::Giga => 1_000_000_000,
+            Self::Gibi => 1 << 30,
+            Self::Tera => 1_000_000_000_000,
+            Self::Tebi => 1 << 40,
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum TimePrefixes {
```

Comment:
Yeah, originally I wasn't going to include minutes and hours since they're different units, but then figured I'd add them even though they don't really fit, which meant adding the suffix to the prefixes. I have an implementation that splits it up into "prefixes", "units", and "values" (values have units, and units may or may not have prefixes), but I feel that it might be too confusing. For now I'll just add comments.